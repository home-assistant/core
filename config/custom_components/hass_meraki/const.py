"""Constants for the Meraki integration."""

DOMAIN = "hass_meraki"
