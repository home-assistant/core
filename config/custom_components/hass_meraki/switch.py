"""The switch Component."""

import logging

import meraki

from homeassistant.components.switch import SwitchEntity
import homeassistant.helpers.device_registry as dr
from homeassistant.helpers.update_coordinator import CoordinatorEntity

from . import DOMAIN

_LOGGER = logging.getLogger(__name__)


async def async_setup_entry(hass, config_entry, async_add_entities):
    """Set up Meraki sensors via config entry."""
    coordinator = hass.data[DOMAIN]["coordinator"]
    device_registry = dr.async_get(hass)
    config_entry = hass.data[DOMAIN].get("config_entry")
    if not config_entry:
        _LOGGER.error("Config entry not found!")
        return

    entities = []
    # Erstelle Sensoren für Geräte, die Client-Daten unterstützen (z. B. Switch, Access Point, Appliance)
    for serial, device_data in coordinator.data.items():
        # Registriere das Gerät im Device Registry
        device_registry.async_get_or_create(
            config_entry_id=config_entry.entry_id,
            identifiers={(DOMAIN, serial)},
            manufacturer="Cisco Meraki",
            name=device_data.get("name", f"Meraki Device {serial}"),
            model=device_data.get("model", "Unknown"),
            sw_version=device_data.get("firmware", "Unknown"),
            connections={("mac", device_data.get("mac", "Unknown"))},
        )

        for port in device_data.get("ports"):
            if port["poe"]["isAllocated"]:
                entities.append(MerakiPoeSwitch(coordinator, serial, port))  # noqa: PERF401
    async_add_entities(entities)

    # Optionally, subscribe to updates to add or remove entities dynamically.
    # Dynamic entity management can be more involved—this is a simplified approach.


class MerakiPoeSwitch(CoordinatorEntity, SwitchEntity):  # noqa: D101
    def __init__(self, coordinator, serial, port):
        """Initialize the switch entity for a specific port."""
        device_data = coordinator.data.get(serial, {})
        super().__init__(coordinator)
        self._serial = serial
        if "lldp" in port:
            self._attr_name = f"Port {port['portId']} {port['lldp']['systemName']}"
        else:
            self._attr_name = f"Port {port['portId']}"
        self._attr_unique_id = f"{serial}_poe_port_{port['portId']}"
        self._attr_has_entity_name = True
        self._port_id = port["portId"]
        self._enabled = port["enabled"]

        # Gerätedaten, damit die Entität dem Gerät zugeordnet wird
        self._attr_device_info = {
            "identifiers": {(DOMAIN, serial)},
            "name": device_data.get("name", f"Meraki Device {serial}"),
            "manufacturer": "Cisco Meraki",
            "model": device_data.get("model", "Unknown"),
            "sw_version": device_data.get("firmware", "Unknown"),
            "connections": {("mac", device_data.get("mac", "Unknown"))},
        }

    @property
    def is_on(self):
        """Return True if this port is enabled."""
        # Assume the API returns a boolean 'enabled' flag
        return self._enabled

    async def async_turn_on(self, **kwargs):
        """Enable this port."""
        result = await self.hass.async_add_executor_job(self._set_poe, True)
        if result:
            self._enabled = result.get("enabled")
            self.async_write_ha_state()

    async def async_turn_off(self, **kwargs):
        """Disable this port."""
        result = await self.hass.async_add_executor_job(self._set_poe, False)
        if result:
            self._enabled = result.get("enabled")
            self.async_write_ha_state()

    def _set_poe(self, enable):
        """Call the Meraki API to change the state for this port."""
        # Adjust the API call to match your needs. This is an example.

        api_key = self.coordinator.config_entry.data["api_key"]
        dashboard = meraki.DashboardAPI(api_key=api_key, suppress_logging=True)
        try:
            # This is a placeholder; replace with the actual API method and parameters.
            return dashboard.switch.updateDeviceSwitchPort(
                self._serial, self._port_id, enabled=enable
            )
        except meraki.APIError as err:
            _LOGGER.error("Failed to set state on port %s: %s", self._port_id, err)
            return False

    # @property
    # def extra_state_attributes(self):
    #     """Optional: expose additional port info as attributes."""
    #     return {"port_name": self._port.get("name")}
