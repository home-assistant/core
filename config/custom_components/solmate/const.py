"""Constants for the solmate integration."""

DOMAIN = "solmate"
