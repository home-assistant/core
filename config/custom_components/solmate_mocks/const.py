"""Constants for the solmate_mocks integration."""

DOMAIN = "solmate_mocks"
