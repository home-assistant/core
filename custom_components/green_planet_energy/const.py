"""Constants for the Green Planet Energy integration."""

DOMAIN = "green_planet_energy"

# Config flow keys
CONF_API_URL = "api_url"

# Default values
DEFAULT_API_URL = "https://mein.green-planet-energy.de/p2"
DEFAULT_SCAN_INTERVAL = 60  # minutes

# API constants
API_METHOD = "getVerbrauchspreisUndWindsignal"
AGGREGATS_ZEITRAUM = ""
AGGREGATS_TYP = ""
SOURCE = "Portal"

# Sensor keys for each hour (9-18)
SENSOR_HOURS = [9, 10, 11, 12, 13, 14, 15, 16, 17, 18]
