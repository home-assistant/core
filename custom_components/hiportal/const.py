"""Constants for hiportal."""

DOMAIN = "hiportal"

CONF_USERNAME = "username"
CONF_PASSWORD = "password"
