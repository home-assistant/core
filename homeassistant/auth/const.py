"""Constants for the auth module."""

from datetime import timedelta

ACCESS_TOKEN_EXPIRATION = timedelta(minutes=30)
MFA_SESSION_EXPIRATION = timedelta(minutes=5)
REFRESH_TOKEN_EXPIRATION = timedelta(days=90).total_seconds()

GROUP_ID_ADMIN = "system-admin"
GROUP_ID_USER = "system-users"
GROUP_ID_READ_ONLY = "system-read-only"
