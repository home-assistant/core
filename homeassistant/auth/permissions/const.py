"""Permission constants."""

CAT_ENTITIES = "entities"
CAT_CONFIG_ENTRIES = "config_entries"
SUBCAT_ALL = "all"

POLICY_READ = "read"
POLICY_CONTROL = "control"
POLICY_EDIT = "edit"
