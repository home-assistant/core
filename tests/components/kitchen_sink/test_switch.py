"""The tests for the demo switch component."""

from collections.abc import Generator
from unittest.mock import patch

import pytest
from syrupy.assertion import SnapshotAssertion

from homeassistant.components.kitchen_sink import DOMAIN
from homeassistant.components.switch import (
    DOMAIN as SWITCH_DOMAIN,
    SERVICE_TURN_OFF,
    SERVICE_TURN_ON,
)
from homeassistant.const import ATTR_ENTITY_ID, STATE_OFF, STATE_ON, Platform
from homeassistant.core import HomeAssistant
from homeassistant.helpers import device_registry as dr, entity_registry as er
from homeassistant.setup import async_setup_component

SWITCH_ENTITY_IDS = ["switch.outlet_1", "switch.outlet_2"]


@pytest.fixture
def switch_only() -> Generator[None]:
    """Enable only the switch platform."""
    with patch(
        "homeassistant.components.kitchen_sink.COMPONENTS_WITH_DEMO_PLATFORM",
        [Platform.SWITCH],
    ):
        yield


@pytest.fixture(autouse=True)
async def setup_comp(hass: HomeAssistant, switch_only: None) -> None:
    """Set up demo component."""
    assert await async_setup_component(hass, DOMAIN, {DOMAIN: {}})
    await hass.async_block_till_done()


async def test_state(
    hass: HomeAssistant,
    device_registry: dr.DeviceRegistry,
    entity_registry: er.EntityRegistry,
    snapshot: SnapshotAssertion,
) -> None:
    """Test switch state."""
    for entity_id in SWITCH_ENTITY_IDS:
        state = hass.states.get(entity_id)
        assert state == snapshot
        entity_entry = entity_registry.async_get(entity_id)
        assert entity_entry == snapshot
        sub_device_entry = device_registry.async_get(entity_entry.device_id)
        assert sub_device_entry == snapshot
        main_device_entry = device_registry.async_get(sub_device_entry.via_device_id)
        assert main_device_entry == snapshot


@pytest.mark.parametrize("switch_entity_id", SWITCH_ENTITY_IDS)
async def test_turn_on(hass: HomeAssistant, switch_entity_id: str) -> None:
    """Test switch turn on method."""
    await hass.services.async_call(
        SWITCH_DOMAIN,
        SERVICE_TURN_OFF,
        {ATTR_ENTITY_ID: switch_entity_id},
        blocking=True,
    )

    state = hass.states.get(switch_entity_id)
    assert state.state == STATE_OFF

    await hass.services.async_call(
        SWITCH_DOMAIN,
        SERVICE_TURN_ON,
        {ATTR_ENTITY_ID: switch_entity_id},
        blocking=True,
    )

    state = hass.states.get(switch_entity_id)
    assert state.state == STATE_ON


@pytest.mark.parametrize("switch_entity_id", SWITCH_ENTITY_IDS)
async def test_turn_off(hass: HomeAssistant, switch_entity_id: str) -> None:
    """Test switch turn off method."""
    await hass.services.async_call(
        SWITCH_DOMAIN,
        SERVICE_TURN_ON,
        {ATTR_ENTITY_ID: switch_entity_id},
        blocking=True,
    )

    state = hass.states.get(switch_entity_id)
    assert state.state == STATE_ON

    await hass.services.async_call(
        SWITCH_DOMAIN,
        SERVICE_TURN_OFF,
        {ATTR_ENTITY_ID: switch_entity_id},
        blocking=True,
    )

    state = hass.states.get(switch_entity_id)
    assert state.state == STATE_OFF
