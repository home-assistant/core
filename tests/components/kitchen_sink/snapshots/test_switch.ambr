# serializer version: 1
# name: test_state
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'Outlet 1',
    }),
    'context': <ANY>,
    'entity_id': 'switch.outlet_1',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'off',
  })
# ---
# name: test_state.1
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'switch',
    'entity_category': None,
    'entity_id': 'switch.outlet_1',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': None,
    'original_icon': None,
    'original_name': None,
    'platform': 'kitchen_sink',
    'previous_unique_id': None,
    'supported_features': 0,
    'translation_key': None,
    'unique_id': 'outlet_1',
    'unit_of_measurement': None,
  })
# ---
# name: test_state.2
  DeviceRegistryEntrySnapshot({
    'area_id': None,
    'config_entries': <ANY>,
    'configuration_url': None,
    'connections': set({
    }),
    'disabled_by': None,
    'entry_type': None,
    'hw_version': None,
    'id': <ANY>,
    'identifiers': set({
      tuple(
        'kitchen_sink',
        'outlet_1',
      ),
    }),
    'is_new': False,
    'labels': set({
    }),
    'manufacturer': None,
    'model': None,
    'model_id': None,
    'name': 'Outlet 1',
    'name_by_user': None,
    'primary_config_entry': <ANY>,
    'serial_number': None,
    'suggested_area': None,
    'sw_version': None,
    'via_device_id': <ANY>,
  })
# ---
# name: test_state.3
  DeviceRegistryEntrySnapshot({
    'area_id': None,
    'config_entries': <ANY>,
    'configuration_url': None,
    'connections': set({
    }),
    'disabled_by': None,
    'entry_type': None,
    'hw_version': None,
    'id': <ANY>,
    'identifiers': set({
      tuple(
        'kitchen_sink',
        '2_ch_power_strip',
      ),
    }),
    'is_new': False,
    'labels': set({
    }),
    'manufacturer': None,
    'model': None,
    'model_id': None,
    'name': 'Power strip with 2 sockets',
    'name_by_user': None,
    'primary_config_entry': <ANY>,
    'serial_number': None,
    'suggested_area': None,
    'sw_version': None,
    'via_device_id': None,
  })
# ---
# name: test_state.4
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'Outlet 2',
    }),
    'context': <ANY>,
    'entity_id': 'switch.outlet_2',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'on',
  })
# ---
# name: test_state.5
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'switch',
    'entity_category': None,
    'entity_id': 'switch.outlet_2',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': None,
    'original_icon': None,
    'original_name': None,
    'platform': 'kitchen_sink',
    'previous_unique_id': None,
    'supported_features': 0,
    'translation_key': None,
    'unique_id': 'outlet_2',
    'unit_of_measurement': None,
  })
# ---
# name: test_state.6
  DeviceRegistryEntrySnapshot({
    'area_id': None,
    'config_entries': <ANY>,
    'configuration_url': None,
    'connections': set({
    }),
    'disabled_by': None,
    'entry_type': None,
    'hw_version': None,
    'id': <ANY>,
    'identifiers': set({
      tuple(
        'kitchen_sink',
        'outlet_2',
      ),
    }),
    'is_new': False,
    'labels': set({
    }),
    'manufacturer': None,
    'model': None,
    'model_id': None,
    'name': 'Outlet 2',
    'name_by_user': None,
    'primary_config_entry': <ANY>,
    'serial_number': None,
    'suggested_area': None,
    'sw_version': None,
    'via_device_id': <ANY>,
  })
# ---
# name: test_state.7
  DeviceRegistryEntrySnapshot({
    'area_id': None,
    'config_entries': <ANY>,
    'configuration_url': None,
    'connections': set({
    }),
    'disabled_by': None,
    'entry_type': None,
    'hw_version': None,
    'id': <ANY>,
    'identifiers': set({
      tuple(
        'kitchen_sink',
        '2_ch_power_strip',
      ),
    }),
    'is_new': False,
    'labels': set({
    }),
    'manufacturer': None,
    'model': None,
    'model_id': None,
    'name': 'Power strip with 2 sockets',
    'name_by_user': None,
    'primary_config_entry': <ANY>,
    'serial_number': None,
    'suggested_area': None,
    'sw_version': None,
    'via_device_id': None,
  })
# ---
